src/DomainGuest.sol

```diff
@@ -0,0 +1,381 @@
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        daiJoin = DaiJoinLike(_daiJoin);
+        vat = daiJoin.vat();
+        dai = daiJoin.dai();
+        claimToken = TokenLike(_claimToken);
+        router = RouterLike(_router);
+
+        vat.hope(_daiJoin);
+        dai.approve(_daiJoin, type(uint256).max);
+        dai.approve(_router, type(uint256).max);
+
+        live = 1;
+    }
+
+    // --- Math ---
+    function _int256(uint256 x) internal pure returns (int256 y) {
+        require((y = int256(x)) >= 0, ARITHMETIC_ERROR);
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+    function _max(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x >= y ? x : y;
+    }
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        unchecked {
+            z = x != 0 ? ((x - 1) / y) + 1 : 0;
+        }
+    }
+
+    // --- Administration ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "end") end = EndLike(data);
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "dust") dust = data;
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // --- Canonical DAI Support ---
+
+    /// @notice Mint DAI and send to user
+    /// @param to The address to send the DAI to on the local domain
+    /// @param amount The amount of DAI to send [WAD]
+    function deposit(address to, uint256 amount) external hostOnly {
+        vat.swell(address(this), _int256(amount * RAY));
+        daiJoin.exit(to, amount);
+
+        emit Deposit(to, amount);
+    }
+
+    /// @notice Withdraw DAI by burning local canonical DAI
+    /// @param to The address to send the DAI to on the remote domain
+    /// @param amount The amount of DAI to withdraw [WAD]
+    function _withdraw(address to, uint256 amount) internal returns (bytes memory payload) {
+        require(dai.transferFrom(msg.sender, address(this), amount), "DomainGuest/transfer-failed");
+        daiJoin.join(address(this), amount);
+        vat.swell(address(this), -_int256(amount * RAY));
+
+        payload = abi.encodeWithSelector(DomainHostLike.withdraw.selector, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    // --- MCD Support ---
+
+    /// @notice Record changes in line and grain and update dss global debt ceiling if necessary
+    /// @param _lid Local ordering id
+    /// @param dline The change in the line [RAD]
+    function lift(uint256 _lid, int256 dline) external hostOnly isLive ordered(_lid) {
```

Comment:
Is there a reason to use `dline` instead of `newLine` as argument here?  `newLine` would be cleaner imo as accidentally issuing (or relaying) twice the `lift` message would not cause problems.

Reply:
I originally chose a relative value before the ordering enforcement was added. This was to allow `lift`s to be commutative. Afterwards it was decided to do ordering enforcement via the `ordered` modifier, but I didn't change it back to an absolute value.

Functionally it will be the same as the `ordered` modifier also enforces at most once execution. For cleanliness I agree that switching to an absolute value is probably better.